spacy_llm/tasks/entity_linker/util.py

```diff
@@ -0,0 +1,144 @@
+    reasons: Optional[List[str]]
+
+    @classmethod
+    def generate(cls, example: Example, **kwargs) -> Optional[Self]:
+        # Check whether all entities have their knowledge base IDs set.
+        n_ents = len(example.reference.ents)
+        n_set_kb_ids = sum([ent.kb_id != 0 for ent in example.reference.ents])
+        if n_ents and n_ents != n_set_kb_ids:
+            warnings.warn(
+                f"Not all entities in this document have their knowledge base IDs set ({n_set_kb_ids} out of "
+                f"{n_ents}). Ignoring {n_set_kb_ids - n_ents} entities in example:\n{example.reference}"
+            )
+        example.reference.ents = [
+            ent for ent in example.reference.ents if ent.kb_id != 0
+        ]
+        if len(example.reference.ents) == 0:
+            return None
+
+        # Assemble example.
+        mentions = [ent.text for ent in example.reference.ents]
+        # Fetch candidates. If true entity not among candidates: fetch description separately and add manually.
+        cands_ents, solutions = kwargs["fetch_entity_info"](example.reference)
+        # If we are to use available docs as examples, they have to have KB IDs set and hence available solutions.
+        assert all([sol is not None for sol in solutions])
+
+        return EntLinkExample(
+            text=EntityLinkerTask.highlight_ents_in_text(example.reference).text,
+            mentions_str=", ".join([f"*{mention}*" for mention in mentions]),
+            mentions=mentions,
+            entity_descriptions=[
+                [cand_ent.description for cand_ent in cands_ent]
+                for cands_ent in cands_ents
+            ],
+            entity_ids=[
+                [cand_ent.id for cand_ent in cands_ent] for cands_ent in cands_ents
+            ],
+            solutions=solutions,
+            reasons=[""] * len(mentions),
+        )
+
+
+def score(examples: Iterable[Example], **kwargs) -> Dict[str, Any]:
+    """Score entity linking accuracy in examples.
+    examples (Iterable[Example]): Examples to score.
+    RETURNS (Dict[str, Any]): Dict with metric name -> score.
+    """
+    return Scorer.score_links(examples, negative_labels=[EntityLinker.NIL])
+
+
+class SpaCyPipelineCandidateSelector:
+    """Callable generated by loading and wrapping a spaCy pipeline with an NEL component and a filled knowledge base."""
+
+    def __init__(
+        self, nlp_path: Union[Path, str], desc_path: Union[Path, str], top_n: int = 5
+    ):
+        """
+        Loads spaCy pipeline, knowledge base, entity descriptions.
+        top_n (int): Top n candidates to include in prompt.
+        nlp_path (Union[Path, str]): Path to stored spaCy pipeline.
+        desc_path (Union[Path, str]): Path to .csv file with descriptions for entities. Has to have two columns
+          with the first one being the entity ID, the second one being the description. The entity ID has to match with
+          the entity ID in the stored knowledge base.
```

Comment:
Do we want to add the decription reader into this functionality? I'm wondering whether we couldn't also make it a separate function that would be easily adjustable by users. Really we just need a callable that transforms entity IDs to descriptions, however they'd like to implement that. Our default could then be this CSV reader thing.

Reply:
Done in https://github.com/explosion/spacy-llm/pull/241/commits/2f7192a033157e2d6fc48262e831fd1cbdea0530.